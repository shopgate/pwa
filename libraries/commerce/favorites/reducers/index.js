import { combineReducers } from 'redux';
import products from './products';
import lists from './lists';

export default combineReducers({
  products,
  lists,
});
