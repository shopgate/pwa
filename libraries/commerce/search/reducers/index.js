import {
  REQUEST_SEARCH_SUGGESTIONS,
  RECEIVE_SEARCH_SUGGESTIONS,
  SEARCH_SUGGESTIONS_LIFETIME,
  SET_SEARCH_SUGGESTIONS_PHRASE,
} from '../constants';

const initialState = {
  suggestionsPhrase: '',
  suggestions: {},
};

/**
 * The search suggestions reducer.
 * @param {Object} [state={}] The current search suggestions state.
 * @param {Object} action The action object.
 * @return {Object} The updated search suggestions data.
 */
const search = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_SEARCH_SUGGESTIONS:
      return {
        ...state,
        suggestions: {
          ...state.suggestions,
          [action.searchPhrase]: {
            isFetching: true,
            suggestions: [],
            expires: 0,
          },
        },
      };

    case RECEIVE_SEARCH_SUGGESTIONS:
      return {
        ...state,
        suggestions: {
          ...state.suggestions,
          [action.searchPhrase]: {
            isFetching: false,
            suggestions: action.suggestions,
            expires: Date.now() + SEARCH_SUGGESTIONS_LIFETIME,
          },
        },
      };
    case SET_SEARCH_SUGGESTIONS_PHRASE:
      return {
        ...state,
        suggestionsPhrase: action.suggestionsPhrase,
      };
    default:
      return state;
  }
};

export default search;
