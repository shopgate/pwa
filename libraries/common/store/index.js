import { composeWithDevTools } from 'redux-devtools-extension';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistState } from '@virtuous/redux-persister';
import benchmarkMiddleware from '@shopgate/pwa-benchmark/profilers/redux';
import benchmarkController from '@shopgate/pwa-benchmark';
import persistedReducers from '../collections/PersistedReducers';
import initSubscribers from '../subscriptions';
import appConfig from '../helpers/config';
import streams from './middelwares/streams';
import logger from './middelwares/logger';

/**
 * The current version of the state created by this reducer.
 * @type {string}
 */
const STATE_VERSION = 'v1';
const storeKey = `shopgate-connect_${STATE_VERSION}`;

/**
 * Returns a normalised initialState from the localstorage.
 * @returns {Object}
 */
function getInitialState() {
  if (!window.localStorage) {
    return undefined;
  }

  const storedState = window.localStorage.getItem(storeKey);

  if (!storedState) {
    return undefined;
  }

  const normalisedState = storedState.replace(new RegExp('"isFetching":true', 'g'), '"isFetching":false');
  return JSON.parse(normalisedState);
}

/**
 * Configures the redux store with all it's middleware and enhancers.
 * @param {Function} reducers The reducers from the theme.
 * @param {Array} subscribers The subscribers to the streams middleware.
 * @return {Object} The redux store.
 */
export function configureStore(reducers, subscribers) {
  // Starts benchmark controller BEFORE adding the middleware.
  if (appConfig.benchmark) {
    benchmarkController.startup();
  }

  const store = createStore(
    reducers,
    getInitialState(),
    composeWithDevTools(
      applyMiddleware(...[
        thunk,
        ...appConfig.benchmark ? [benchmarkMiddleware] : [],
        streams,
        logger,
      ]),
      persistState({
        key: storeKey,
        paths: persistedReducers.getAll(),
      })
    )
  );

  initSubscribers(subscribers);

  if (window.Cypress) {
    window.store = store;
  }

  return store;
}

/**
 * Configures a mocked store for tests.
 * @param {Function} reducers The reducers to use in the mock store.
 * @param {Array} subscribers Streams subscribers to initialize.
 * @returns {Store}
 */
export function createMockStore(reducers = () => {}, subscribers = []) {
  const store = createStore(
    reducers,
    undefined,
    applyMiddleware(thunk, streams)
  );

  if (subscribers.length !== 0) {
    initSubscribers([].concat(subscribers));
  }

  return store;
}
