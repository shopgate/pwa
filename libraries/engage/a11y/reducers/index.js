import { produce } from 'immer';
import { isDev, logger } from '@shopgate/engage/core/helpers';
import { A11Y_INCREASE_MODAL_COUNT, A11Y_DECREASE_MODAL_COUNT } from '../constants';

const defaultState = {
  modalCount: 0,
};

/**
 * Stores a11y state
 * @param {Object} [state={}] The current state.
 * @param {Object} action The action object.
 * @returns {Object} The new state.
 */
export default function a11y(state = defaultState, action) {
  /* eslint-disable no-param-reassign */
  const producer = produce((draft) => {
    switch (action.type) {
      case A11Y_INCREASE_MODAL_COUNT: {
        draft.modalCount += 1;
        break;
      }
      case A11Y_DECREASE_MODAL_COUNT: {
        draft.modalCount -= 1;
        if (draft.modalCount < 0) {
          if (isDev) {
            logger.warn('Modal count is negative, resetting to 0');
          }
          draft.modalCount = 0;
        }
        break;
      }
      default:
        break;
    }
  });
  /* eslint-enable no-param-reassign */

  return producer(state);
}
