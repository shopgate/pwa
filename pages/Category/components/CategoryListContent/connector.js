import { connect } from 'react-redux';
import isEqual from 'lodash/isEqual';
import {
  getChildCategoriesById,
  getCurrentCategoryChildCount,
  hasChildren,
} from '@shopgate/pwa-common-commerce/category/selectors';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @param {Object} props The component props.
 * @return {Object} The extended component props.
 */
const mapStateToProps = (state, props) => ({
  categories: getChildCategoriesById(state, props),
  hasChildren: hasChildren(state, props),
  childrenCount: getCurrentCategoryChildCount(state, props),
});

/**
 * @param {Object} next The next component props.
 * @param {Object} prev The previous component props.
 * @returns {boolean}
 */
const areStatePropsEqual = (next, prev) => {
  if (prev.hasChildren !== next.hasChildren) {
    return false;
  }

  if (!isEqual(prev.categories, next.categories)) {
    return false;
  }

  if (prev.childrenCount !== next.childrenCount) {
    return false;
  }

  return true;
};

export default connect(mapStateToProps, null, null, { areStatePropsEqual });
