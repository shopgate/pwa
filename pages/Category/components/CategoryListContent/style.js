import { css } from 'glamor';

export const noPaddingTop = css({
  paddingTop: 0,
}).toString();

