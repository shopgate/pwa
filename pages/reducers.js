import { combineReducers } from 'redux';
import {
  persistedReducers,
  configuration,
  RESET_APP_REDUCERS,
  hasWebBridge,
} from '@shopgate/engage/core';
import checkout from '@shopgate/engage/checkout/reducers';
import client from '@shopgate/pwa-common/reducers/client';
import url from '@shopgate/pwa-common/reducers/url';
import user from '@shopgate/pwa-common/reducers/user';
import page from '@shopgate/pwa-common/reducers/page';
import router from '@shopgate/pwa-common/reducers/router';
import menu from '@shopgate/pwa-common/reducers/menu';
import modal from '@shopgate/pwa-common/reducers/modal';
import cart from '@shopgate/pwa-common-commerce/cart/reducers';
import category from '@shopgate/pwa-common-commerce/category/reducers';
import favorites from '@shopgate/pwa-common-commerce/favorites/reducers';
import filter from '@shopgate/pwa-common-commerce/filter/reducers';
import product from '@shopgate/pwa-common-commerce/product/reducers';
import { settings } from '@shopgate/engage/core/reducers';
import locations from '@shopgate/engage/locations/reducers';
import orders from '@shopgate/engage/orders/reducers';
import search from '@shopgate/pwa-common-commerce/search/reducers';
import reviews from '@shopgate/pwa-common-commerce/reviews/reducers';
import extensions from 'Extensions/reducers';

persistedReducers.set([
  'cart',
  'client.info',
  'page',
  'locations.storage',
  'locations.userFormInput',
  'url',
  'user',
  ...(hasWebBridge() ? 'menu' : []),
]);

configuration.set(RESET_APP_REDUCERS, [
  'cart',
  'category',
  'product',
  'favorites',
  'filter',
  'reviews',
  'search',
]);

const reducers = combineReducers({
  router,
  cart,
  category,
  checkout,
  client,
  ...extensions && { extensions: combineReducers(extensions) },
  favorites,
  filter,
  menu,
  modal,
  page,
  product,
  locations,
  orders,
  settings,
  reviews,
  search,
  url,
  user,
});

export default reducers;
