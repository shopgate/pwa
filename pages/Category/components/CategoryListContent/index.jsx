import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import Portal from '@shopgate/pwa-common/components/Portal';
import {
  CATEGORY_LIST,
  CATEGORY_LIST_AFTER,
  CATEGORY_LIST_BEFORE,
} from '@shopgate/pwa-common-commerce/category/constants/Portals';
import { Section } from '@shopgate/engage/a11y';
import { ResponsiveContainer } from '@shopgate/engage/components';
import CategoryList from 'Components/CategoryList';
import CategoryGrid from 'Components/CategoryGrid';
import connect from './connector';

/**
 * The category list content.
 */
class CategoryListContent extends PureComponent {
  static propTypes = {
    categoryId: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.shape()),
    childrenCount: PropTypes.number,
    hasChildren: PropTypes.bool,
  };

  static defaultProps = {
    categories: null,
    childrenCount: 6,
    hasChildren: false,
  }

  /**
   * @returns {JSX}
   */
  render() {
    const {
      hasChildren, categories, categoryId, childrenCount,
    } = this.props;

    return (
      <Fragment>
        <Portal name={CATEGORY_LIST_BEFORE} props={{ categoryId }} />
        <Portal name={CATEGORY_LIST} props={{ categoryId }}>
          {hasChildren && (
            <Section title="category.sections.categories">
              <ResponsiveContainer appAlways breakpoint="<=xs">
                <CategoryList categories={categories} prerender={childrenCount} />
              </ResponsiveContainer>
              <ResponsiveContainer webOnly breakpoint=">xs">
                <CategoryGrid categories={categories} prerender={childrenCount} />
              </ResponsiveContainer>
            </Section>
          )}
        </Portal>
        <Portal name={CATEGORY_LIST_AFTER} props={{ categoryId }} />
      </Fragment>
    );
  }
}

export default connect(CategoryListContent);
