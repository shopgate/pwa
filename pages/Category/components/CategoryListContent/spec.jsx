import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { createMockStore } from '@shopgate/pwa-common/store';
import reducers from 'Pages/reducers';
import CategoryListContent from './index';

const store = createMockStore(reducers);
jest.mock('../../../../components/View/context.js');

describe('<CategoryListContent />', () => {
  it('should render', () => {
    const wrapper = mount((
      <Provider store={store}>
        <CategoryListContent categoryId="1234" />
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
