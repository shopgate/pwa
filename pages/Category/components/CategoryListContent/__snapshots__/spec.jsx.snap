// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryListContent /> should render 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectFunction
    categoryId="1234"
  >
    <CategoryListContent
      categories={null}
      categoryId="1234"
      childrenCount={null}
      dispatch={[Function]}
      hasChildren={false}
    >
      <Portal
        name="category.list.before"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
      <Portal
        name="category.list"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
      <Portal
        name="category.list.after"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
    </CategoryListContent>
  </ConnectFunction>
</Provider>
`;
